Add option to skip PublishSidebar on publishing

oandregal opened the pull request:
This PR adds a new option to skip the PublishSidebar on publishing.

### How it works

The option is surfaced both in the "Tools" section within the general menu and the "Publish sidebar" itself:

![peek 2018-09-11 20-24](https://user-images.githubusercontent.com/583546/45379653-f5821b80-b600-11e8-8cf5-eb4364f1efa3.gif)

This is how it looks when the option is enabled:

![peek 2018-09-11 20-25](https://user-images.githubusercontent.com/583546/45379667-fd41c000-b600-11e8-90a5-ea0864140e8c.gif)

This is how it looks when the option is disabled:

![peek 2018-09-11 20-26](https://user-images.githubusercontent.com/583546/45379678-03d03780-b601-11e8-9d57-930aab0c42f8.gif)

### Testing

- Open a new post
- Add some content
- Click "Publish..." button (which should show three dots at the end).

The expected result is that the Pre-Publish panel is opened and a checkbox is shown at the bottom of it.

- Disable the checkbox, and cancel publishing.
- Click "Publish" button again (which shouldn't have the dots).

The expected result is that the post is published directly, without the pre-publish checks.

These can also be disabled from "Tools" section of the general sidebar. Play a bit with those checks and make sure it works as expected.

### TODO 

- [x] Add "Don't show this sidebar again" checkbox to the Publish Sidebar.
- [x] Better style the dismissable checkbox.
- [x] Move preference logic to `core/editor`.
- [x] Update tests.

oandregal committed:
Add PrePublish panel option to the general menu tools section

oandregal committed:
Add data API

oandregal committed:
Rename PublishPanel to PrePublishPanel

oandregal committed:
Implement core/edit-post actions and selector

oandregal committed:
Rename pre-publish-panel-toggle to publish-sidebar-toggle

oandregal committed:
Connect the dots: publish toggles publish sidebar or publish directly

oandregal committed:
Update test for isPublishSidebarEnable preference

oandregal committed:
Expose dismissing option as a checkbox in the sidebar

oandregal committed:
Move publish sidebar logic to core/editor store

This logic is used by the PostPublishPanel, which lives in editor.
We don't want editor to depend on anything from edit-post.

An alternative approach could be to keep the logic in edit-post
and use the the PrePublish and PostPublish extensions.
In my view, those are meant for allowing plugins to inject their
own panels into the PublishPanel. Dismissing the panel is a core
action and using the extensions to circumvent the editor/edit-post
separation would be a convoluted way to add support for dismissing the panel.

nosolosw commented on packages/editor/src/store/actions.js:
```diff
@@ -763,3 +763,25 @@ export function unregisterToken( name ) {
+ *
+ * @return {Object} Action object
+ */
+export function enablePublishSidebar() {
```
I originally added the enable/disable logic to the `core/edit-post` store and moved it later to the `core/editor`. See 28262d227182646cc0c5bc85d69999f16e5df2e4 The rationale for this change is that the PostPublishPanel (which lives in `@wordpress/editor`) uses this logic, and the `editor` can't depend on any component declared in `edit-post`.

An alternative approach could have been to keep the logic in `edit-post` and use the PrePublish and PostPublish extensions from `edit-post` to inject the checkbox component into the Publish sidebar. This is why I've discarded this approach:

- In my view, the extensions are meant for allowing plugins to inject their own panels into the PublishPanel. Dismissing the panel is a core action and using the extensions to circumvent the editor/edit-post separation would be a convoluted way to add support for it. To me, that'd be a smell that the editor/edit-post separation needs some more thinking.
- We want the checkbox to be shown in the same position of the panel (the bottom). To achieve this, we'd need to rely on CSS styling, because a plugin can't control other plugins's positions in the panel - which is prone to conflicts and unreliable positioning the checkbox.

aduth replied:
Is there another alternative to consider where the publish panel makes more sense in `edit-post` and could be moved wholesale there? This doesn't really feel like a component which ought be offered by the `editor` module.

aduth replied:
I wish we'd taken a more generalized approach akin to `edit-post`'s `toggleFeature` than to have specific action creators, selectors, reducer handling for each and every user preference.

tofumatt commented on edit-post/components/header/more-menu/index.js:
```diff
@@ -38,6 +39,7 @@ const MoreMenu = () => (
 					filterName="editPost.MoreMenu.tools"
 				>
 					<TipsToggle onToggle={ onClose } />
+					<PublishSidebarToggle onToggle={ onClose } />
```
I know it's unrelated to this PR, but the `onClose` method being used for `onToggle` several times and `onSelect` below makes me think it should be renamed to `closeMenu` or something. 🤷‍♂️

tofumatt commented on edit-post/components/header/publish-button/index.js:
```diff
@@ -0,0 +1,25 @@
+import { PostPublishPanelToggle, PostPublishButton } from '@wordpress/editor';
+
+const PublishButton = function( {
```
Any reason for using `function() {` over `() => {`? We just tend toward fat-arrow functions in the codebase, so it stood out.

nosolosw replied:
4acd0c5086ae52408b5fcfe8a96d6dfe754183cc

tofumatt commented on edit-post/components/header/publish-button/index.js:
```diff
@@ -0,0 +1,25 @@
+	isSidebarEnabled,
+	onToggle,
+	forceIsDirty,
+	forceIsSaving } ) {
```
Odd that eslint didn't catch it, but I'd expect `} ) {` on a de-dented newline.

nosolosw replied:
7322881e6f257dde3ea4312a43b3f068b0ee3e35

tofumatt commented on edit-post/components/header/publish-button/index.js:
```diff
@@ -0,0 +1,25 @@
+import { PostPublishPanelToggle, PostPublishButton } from '@wordpress/editor';
+
+const PublishButton = function( {
```
So this is now a button that turns into a panel? I'd sort of expect we'd have a button that toggled a panel on/off, not really one that changes from panel to button itself...

tofumatt commented on edit-post/components/header/publish-sidebar-toggle/index.js:
```diff
@@ -0,0 +1,37 @@
+const PublishSidebarToggle = function( { onToggle, isEnabled } ) {
+	return (
+		<MenuItem
+			icon={ isEnabled && 'yes' }
```
The icon here should be `'yes'`? That seems odd to me.

nosolosw replied:
That's how the [MenuItem](https://github.com/WordPress/gutenberg/tree/2d1e3e055c31a2df03aec83ce940a5dfe17447c6/packages/components/src/menu-item) works, it uses the [yes dashicon](https://developer.wordpress.org/resource/dashicons/#yes).

sarahmonster replied:
Ideally this could switch to using the Material tick instead of the Dashicons tick, but that's obviously a bigger change that's out-of-scope for this PR.

tofumatt commented on packages/editor/src/components/post-publish-panel/index.js:
```diff
@@ -96,6 +96,11 @@ class PostPublishPanel extends Component {
 						</PostPublishPanelPostpublish>
 					) }
+					<CheckboxControl
+						label={ __( "Don't show this sidebar again." ) }
```
This should be an apostrophe (`’`, not a single-quote `'`).

nosolosw replied:
I need to update my [keyboard setup](http://nosolosoftware.com/touch-typing-in-dvorak/) to include that apostrophe. Good catch!

nosolosw replied:
874a40b064cec0e818b8e7f9b31fd592c039bb66

tofumatt commented on packages/editor/src/store/reducer.js:
```diff
@@ -810,6 +810,13 @@ export function preferences( state = PREFERENCES_DEFAULTS, action ) {
 			};
+
+		case 'ENABLE_PUBLISH_SIDEBAR':
+		case 'DISABLE_PUBLISH_SIDEBAR':
```
I get that this is technically less code but I think it's nicer to have each action type return its own object. It's easier to grok what's happening that way–this code made sense to me but only after I looked at it for probably ten seconds; I was wondering why `'DISABLE_PUBLISH_SIDEBAR'` needed to check if the action was `'ENABLE_PUBLISH_SIDEBAR'`. I think it'd be nicer to just do:

```js
case 'ENABLE_PUBLISH_SIDEBAR':
	return {
		...state,
		isPublishSidebarEnabled: true,
	};
case 'DISABLE_PUBLISH_SIDEBAR':
	return {
		…state,
		isPublishSidebarEnabled: false,
	};
```

nosolosw replied:
Thanks for keeping us obvious.

nosolosw replied:
ee882d43c654291f10cb5e97bc8e09cfd223b14a

tofumatt committed:
docs: tweak comment on isPublishSidebarEnabled

tofumatt reviewed (changes requested):
Since there's still stuff marked todo I didn't check this out and try it yet but I left some comments on the code and approach.

I wonder if it would be better to have the panel appear as a component next to the publish button rather than have the publish button switch between being a button and a panel. It seems a bit odd to place a button and have its component conditionally output one of two very separate things.

I guess if we do it often elsewhere there's precedence, but I don't recall seeing it often in the codebase and I find it a bit less of an explicit/easy-to-follow approach.

Does that make sense?

oandregal committed:
Use arrow function instead of traditional function

oandregal committed:
Fix linting issue that went undetected by eslint

oandregal committed:
Use english apostrophe, not single commas

oandregal committed:
Make enable/disable actions more obvious

oandregal committed:
Inline PostPublish button / toggle logic

sarahmonster reviewed (changes requested):
So! When I tested this out the first time, it defaulted to not-showing the pre-publish panel. I think it would be better if we made the pre-publish panel *opt-out*, rather than opt-in. That is, you can hide it if you want to, but by default, it's shown.

Next! Let's avoid calling it a "sidebar", since a) it's not exactly behaving as a sidebar once all our pre-publish panel changes land in #9398 and b) it's not a sidebar at all on mobile. I keep waffling on what to call it and how to phrase this, so I think it'd be good to put out a batsignal for some editorial help!

Some ideas:
- pre-publish panel
- publish checkup
- publish confirmation

I've realised we should also probably match the checkbox behaviour—so if the checkbox is checked in the tools menu, it should also be checked in the pre-publish panel. That will involve more wording changes, and it might make sense to make it a positive ("show the panel" rather than a negative ("don't show the panel), but it could work either way, so long as we ensure their states match.

In the tools menu:
"Show publish confirmation" (positive)
or "Publish posts immediately" (negative)

And matching statements for the panel checkbox:
"Show this confirmation every time I publish a post" (positive)
or "Don't show this next time I publish a post" (negative)

These would ideally be a bit shorter if possible too. 

Third, I'll make a wee CSS change here that'll fix the margins issue for this PR, even though it'll just be a band-aid until we get all the other changes introduced as well.

oandregal committed:
Add padding to align and give it space

oandregal committed:
Align to bottom

nosolosw commented:
> Third, I'll make a wee CSS change here that'll fix the margins issue for this PR, even though it'll just be a band-aid until we get all the other changes introduced as well.

I was already working on that! Changes already pushed.

oandregal committed:
Update wording

nosolosw commented:
> So! When I tested this out the first time, it defaulted to not-showing the pre-publish panel. I think it would be better if we made the pre-publish panel opt-out, rather than opt-in. That is, you can hide it if you want to, but by default, it's shown.

mmm, I've tried by clearing my browser's cache and deleting any local changes I might have and it works as an opt-in. That was the intention. Perhaps an earlier bug? Would you mind clearing your browser's cache and try again?

Updated the wording and changed the state of checkbox and tools menu item to be equal.

nosolosw commented:
@tofumatt I'll defer to @sarahmonster and @karmatosed about the interaction, as I know they've discussed it and explored options in https://github.com/WordPress/gutenberg/issues/7602 and https://github.com/WordPress/gutenberg/pull/9398

michelleweber commented:
I'd call it "pre-publish check" and I'm thinking the clearest thing is to use to turn it on is the same language as the confirmation: "Double-check settings before publication" -- If that's too long, maybe just "Enable pre-publish checks" -- and "Prompt me to double-check settings every time I publish something."

sarahmonster commented:
"Enable pre-publish checks" is probably best for the toolbar given length, and it's very clear. 💯 

What about making the checkbox in the panel itself mirror this wording more directly—something along the lines of "Show this pre-publish check every time I publish"?

oandregal committed:
Update copy

nosolosw commented:
Updated copy in d7d1b77

![screenshot from 2018-09-12 09-30-58](https://user-images.githubusercontent.com/583546/45409199-d621d780-b66e-11e8-88e5-5e41889f7ccc.png)

![screenshot from 2018-09-12 09-31-32](https://user-images.githubusercontent.com/583546/45409201-d6ba6e00-b66e-11e8-8682-307ab818cf03.png)

sarahmonster committed:
Tweak microcopy for pre-publish checks.

sarahmonster commented:
> Perhaps an earlier bug? Would you mind clearing your browser's cache and try again?

Seems to be working as expected now, or at least I can't replicate anymore, so 👍.

> I wonder if it would be better to have the panel appear as a component next to the publish button rather than have the publish button switch between being a button and a panel. It seems a bit odd to place a button and have its component conditionally output one of two very separate things.

I've tried to parse this several times and can't make heads or tails of it, but if you can clarify @tofumatt  I'm happy to give feedback on the interaction pattern!

tofumatt commented:
Sorry, my comment was just about coding; the interaction would be the same, but I thought it was strange that the component conditionally output two very separate components based on state.

nosolosw commented:
> Sorry, my comment was just about coding; the interaction would be the same, but I thought it was strange that the component conditionally output two very separate components based on state.

Do you refer to https://github.com/WordPress/gutenberg/pull/9760/files#diff-a6f2fb9ba2b6fbe3ad649f4c019b76deR53 right? Conditionally show a button or a toggle depending on whether the pre-publish checks are enabled or not. I'm not sure I understand your concerns. Could we perhaps move to code and see if there is a different implementation? Let's comment there.

tofumatt commented:
@nosolosw Looks like you changed the code and alleviated my concerns 😄

Also: is the code ready for another review? I can look if so 😄

nosolosw commented:
> Also: is the code ready for another review? I can look if so

Please! Also cc @sarahmonster in case she has any design concerns I should address. It'd be lovely to merge this today or tomorrow so it can make it to the 3.9 release.

tofumatt commented on packages/editor/src/components/post-publish-panel/index.js:
```diff
@@ -97,6 +97,13 @@ class PostPublishPanel extends Component {
 				</div>
+				<div className="editor-post-publish-panel__footer">
+					<CheckboxControl
+						label={ __( 'Always show pre-publish checks.' ) }
```
Putting the word "this" inside this text might help contextualise this more.

> Always show these pre-publish checks.

might help users understand it more. If it doesn't look too cramped I say go with that, but if it's already a tight space feel free to leave it as-is.

tofumatt reviewed (changes requested):
Looks good, but when I checked it out and tested it the pre-publish panel wasn't on by default, which I'd think it should be.

I saw that eventually going to another page or something fixed it for @sarahmonster… but as I already think this could use some testing I'd like to see an E2E test that makes sure it's enabled by default on a new site and that it can be disabled before merging.

Code and experience works great though, nice job 👍

tofumatt commented:
> Could you clear the browser cache and test again? May it be that the option was unchecked on a previous session? I'm going to ask people that haven't tested this yet to give a spin as well.

I hadn't checked this branch out before so I'd never set the value on/off. Because we want the value to be on by default for any user after this ships we don't want them to have to enable browser cache to have it enabled…

That said I refreshed an existing page at first and when I reset my preferences it worked… 🤷‍♂️

sarahmonster commented:
I am a-okay with things from a design standpoint, but it definitely sounds like it'd be worth writing a test or two to ensure that nothing breaks in the future. We'll definitely want to be sure it isn't possible to get it accidentally disabled.

If it helps at all, mine self-corrected when I opened things in another browser.

oandregal committed:
Remove outdated docs

These data has been moved to edit-post store as per
https://github.com/WordPress/gutenberg/pull/4661

nosolosw commented:
Found the bug! Here's how to reproduce: 

1. Clear your cache.
2. Check out  the `master` branch and do stuff like saving a post.
3. Check out this branch.

When you open a post, the "Publish" button should publish directly. Which is not what we want.

I haven't caught this before due to the [observer effect](https://simple.wikipedia.org/wiki/Observer_effect). So thanks @tofumatt to prompt me to dig deeper into this. The reason for this is:

> There are only two hard things in Computer Science: cache invalidation and naming things.

We use a `localStorage` mechanism to persist parts of the state in between sessions (like whether the user has already seen the tips or not). In the case of the `core/editor` store we persist the [preferences](https://github.com/WordPress/gutenberg/blob/c1c4c7ed80f0f29f60adb44f2fd0cb2f8660045d/packages/editor/src/store/index.js#L30) key, which contains an object. This is the problem: the `localStorage` mechanism doesn't check whether the cached object and the object in the state have the same shape, it just overwrites the running state (so new keys such as `isPublishSidebarEnabled` are deleted). We may have the same problem in other parts of our codebase: [nux store](https://github.com/WordPress/gutenberg/blob/c1c4c7ed80f0f29f60adb44f2fd0cb2f8660045d/packages/nux/src/store/index.js) and [edit-post store](https://github.com/WordPress/gutenberg/blob/c1c4c7ed80f0f29f60adb44f2fd0cb2f8660045d/edit-post/store/index.js).

I don't think this would be caught by an e2e test as it involves browser cache and past sessions state (the `preferences` object may have been void, for example - in that case everything would work fine).

Need to run some errands in a moment. Will have a fix today or tomorrow.

jasmussen commented:
As part of this PR, be sure to test mobile and conditions where a user can't publish, or when you are scheduling a post, as well as translations. 

There's some context in https://github.com/WordPress/gutenberg/pull/5300, https://github.com/WordPress/gutenberg/pull/5329, and https://github.com/WordPress/gutenberg/pull/6353.

oandregal committed:
Preventive measure

In theory, isPublishSidebarEnabled shouldn't be ever undefined
in the preferences object, so this shouldn't be necessary.

In practice, we use a localStorage mechanism to persist
parts of the state in between sessions (like whether the user
has already seen the tips or not). For the core/editor store
we persist the preferences key, which contains an object.
This is the problem: the localStorage mechanism doesn't check
whether the cached preferences object and the state preferences object
have the same shape, it just overwrites the current state,
so new added keys are deleted.

While this bug is not fixed and deployed,
we want this preventive measure in place.

oandregal committed:
Fix and add tests

nosolosw commented:
@tofumatt I think it makes sense to fix the Redux persistence mechanism as a different PR, it may require wider testing isolated from this specific use case. So, the quick fix specific to this PR is this: make sure the `isPublishSidebarEnabled` selector takes the default value if the key is not in the preferences object.

tofumatt reviewed (commented):
The preventative measure looks good to me. I'd like to see an end-to-end test for the general experience around the pre-publish panel appearing and being dismissible. I think it would be straightforward to write and useful to test. After E2E tests are added I'm cool with the code and the fix on preferences works for me.

oandregal committed:
Add e2e test

oandregal committed:
Refactor code to avoid duplication

nosolosw commented:
Tested scheduling & submit for review states, on mobile, translations, etc.

@tofumatt I can land this as soon as I have a :+1: to enable merging again.

tofumatt committed:
chore: Tweak styles

tofumatt commented on test/e2e/specs/publishing.test.js:
```diff
@@ -4,48 +4,53 @@
+		describe( `a ${ postType } with pre-publish checks disabled`, () => {
+			beforeEach( async () => {
+				await newPost( postType );
+				await disablePrePublishChecks();
```
Because this setting persists between page loads we should be careful with this… arguably we should have some kind of consistent/known state and after each test where we disable the pre-publish check we should re-enable them with the `afterEach` test helper.

Right now this test will disable pre-publish checks elsewhere which–as it's not the default state–could be confusing to future test writers.

nosolosw replied:
Done, good catch!

tofumatt commented on test/e2e/support/utils.js:
```diff
@@ -136,6 +136,12 @@ export async function newPost( { postType, enableTips = false } = {} ) {
+export async function disablePrePublishChecks( ) {
+	await page.click( '.edit-post-more-menu' );
+	await page.waitForSelector( '.components-popover__content' );
+	await page.click( '.components-popover__content > .components-menu-group:nth-child(3) button:nth-child(3)' );
```
This is a bit of a fragile selector, is there anything better/better-named we could use to select the "Don't show this again" checkbox? It looks like `. editor-post-publish-panel__footer input[type=checkbox]` would work assuming `CheckboxControl` outputs an actual checkbox.

nosolosw replied:
Solved at d7c7151

I was reluctant to introduce a specific selector _just for testing_ but, on a second thought, it's totally worth it in this case. Note also that the code uses the general menu, not the actual checkbox within the publish sidebar. We can't guarantee the publish sidebar is enabled unless we use the general menu.

tofumatt reviewed (approved):
I think we need to reset the setting for pre-publish panel being disabled and I left a comment about a selector, but otherwise this looks good.

Once the tests reset the setting so they aren't causing side-effects outside their test runs this is good to merge, so approving conditional on that change 😄

oandregal committed:
Disable pre-publish checks after test

oandregal committed:
Add enable/disable functions

oandregal committed:
Use specific selector to target pre-publish checks menu item

oandregal committed:
Do not make assumptions about previous states

mintplugins commented:
Just to throw another opinion out there, has the problem of having multiple User Experiences been raised as a concern here? If some people experience it one way, and others experience it another way (based on a setting they'll possibly/likely forget about), it creates a disparate UX, making it tougher to write tutorials, make tutorial videos, and even to develop plugins which tap into this area. 

Personally I don't like this area, but I think I like having the option to turn it off even less, as it created a forever-more fragmented user experience. My apologies for being a downer!

aduth commented on packages/editor/src/store/test/reducer.js:
```diff
@@ -1580,9 +1580,28 @@ describe( 'state', () => {
 		} );
 
+		it( 'should disable the publish sidebar', () => {
+			const original = deepFreeze( preferences( undefined, { } ) );
```
>No filler spaces in empty constructs (e.g., {}, [], fn()).

https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#spacing

Need to find if possible to enforce by ESLint.

aduth replied:
https://github.com/eslint/eslint/issues/7927
